Serve packages archived (as zip) on storage

jsoriano opened the pull request:
Draft exploring the idea of serving packages stored directly as zip files, so users receive the same packages as stored in the backend.

An "`Indexer"` interface is introduced to allow to provide different implementations for sources of packages.
List of packages is not global anymore, allowing better isolation of test cases. 

This is done in context of #670.

Checklist:
- [ ] Refactor the artifacts handler so it uses the paths from the indexer.
- [ ] Refactor the indexer so it offers a more efficient way to look for specific packages and versions (probably needed for previous point).
- [ ] Some conditional code is added to decide how to handle packages depending on how they are stored, make it more explicit and safer in general.
- [ ] In the filesystem indexer, collect the `Package` objects from the zip files without extracting them to disk.

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fpackage-registry%2FPR-703/detail/PR-703/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fpackage-registry%2FPR-703/detail/PR-703/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fpackage-registry%2FPR-703/detail/PR-703/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fpackage-registry%2FPR-703/detail/PR-703/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://package-registry_703.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2021-07-30T12:20:25.871Z&rangeTo=2021-07-30T12:40:25.871Z&transactionName=Ingest-manager/package-registry/PR-{number}&transactionType=unknown&latencyAggregationType=avg&traceId=1c657c79bbe4ab9514a9babf01fbe460&transactionId=71f9c9ac78b60927)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2021-07-30T12:30:25.871+0000


* Duration: 3 min 34 sec


* Commit: 9aa0a63786794737c393bdc38f69e5fede82b54f




#### Trends :test_tube:


![Image of Build Times](https://beats-ci.elastic.co/job/Ingest-manager/job/package-registry/job/PR-703/buildTimeGraph/png)



</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->









### Steps errors [![1](https://img.shields.io/badge/1%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fpackage-registry%2FPR-703/detail/PR-703/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Checks formatting / linting`
<ul>
<li>Took 0 min 10 sec . View more details on <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/package-registry/pipelines/PR-703/runs/1/steps/129/log/?start=0">here</a></li>
<li>Description: <code>mage -debug check</code></l1>
</ul>
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2021-07-30T12:32:10.191Z] Checking out Revision 9aa0a63786794737c393bdc38f69e5fede82b54f (PR-703)
[2021-07-30T12:32:10.282Z] Commit message: "Replace global package list with indexer"
[2021-07-30T12:32:10.282Z] Cleaning workspace
[2021-07-30T12:32:10.391Z]  > git --version # timeout=10
[2021-07-30T12:32:10.402Z]  > git --version # 'git version 2.25.1'
[2021-07-30T12:32:10.423Z] fatal: bad object e3247d1cb85270a6b8783853a8449883446c4ec6
[2021-07-30T12:32:10.107Z]  > git config core.sparsecheckout # timeout=10
[2021-07-30T12:32:10.110Z]  > git checkout -f 9aa0a63786794737c393bdc38f69e5fede82b54f # timeout=15
[2021-07-30T12:32:10.169Z]  > git remote # timeout=10
[2021-07-30T12:32:10.172Z]  > git config --get remote.origin.url # timeout=10
[2021-07-30T12:32:10.175Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2021-07-30T12:32:10.178Z]  > git merge ce683f8b857d04c536bf74a1764e32bac07fd3fa # timeout=10
[2021-07-30T12:32:10.186Z]  > git rev-parse HEAD^{commit} # timeout=10
[2021-07-30T12:32:10.194Z]  > git config core.sparsecheckout # timeout=10
[2021-07-30T12:32:10.197Z]  > git checkout -f 9aa0a63786794737c393bdc38f69e5fede82b54f # timeout=15
[2021-07-30T12:32:10.285Z]  > git rev-parse --verify HEAD # timeout=10
[2021-07-30T12:32:10.289Z] Resetting working tree
[2021-07-30T12:32:10.289Z]  > git reset --hard # timeout=10
[2021-07-30T12:32:10.362Z]  > git clean -fdx # timeout=10
[2021-07-30T12:32:11.650Z] Masking supported pattern matches of $GIT_USERNAME or $GIT_PASSWORD
[2021-07-30T12:32:12.185Z] + git fetch https://****:****@github.com/elastic/package-registry.git +refs/pull/*/head:refs/remotes/origin/pr/*
[2021-07-30T12:32:13.656Z] Running in /var/lib/jenkins/workspace/_manager_package-registry_PR-703/src/github.com/elastic/package-registry/.git
[2021-07-30T12:32:13.760Z] Archiving artifacts
[2021-07-30T12:32:14.384Z] + git rev-parse HEAD
[2021-07-30T12:32:14.716Z] + git rev-parse HEAD
[2021-07-30T12:32:15.032Z] + git rev-parse origin/pr/703
[2021-07-30T12:32:15.081Z] [INFO] githubEnv: Found Git Build Cause: pr
[2021-07-30T12:32:15.531Z] Masking supported pattern matches of $GITHUB_TOKEN
[2021-07-30T12:32:16.500Z] [WARN] githubApiCall: The REST API call https://api.github.com/repos/elastic/package-registry/pulls/703/reviews return 0 elements
[2021-07-30T12:32:16.556Z] [INFO] githubPrCheckApproved: Title: Serve packages archived (as zip) on storage - User: jsoriano - Author Association: MEMBER
[2021-07-30T12:32:25.159Z] Stashed 1447 file(s)
[2021-07-30T12:32:25.252Z] Running in /var/lib/jenkins/workspace/_manager_package-registry_PR-703/src/github.com/elastic/package-registry
[2021-07-30T12:32:28.375Z] + gvm 1.15.7
[2021-07-30T12:32:33.652Z] export GOROOT="/var/lib/jenkins/workspace/_manager_package-registry_PR-703/.gvm/versions/go1.15.7.linux.amd64"
[2021-07-30T12:32:33.652Z] export PATH="/var/lib/jenkins/workspace/_manager_package-registry_PR-703/.gvm/versions/go1.15.7.linux.amd64/bin:$PATH"
[2021-07-30T12:32:34.062Z] + go get -u github.com/magefile/mage
[2021-07-30T12:32:38.622Z] + go get -u github.com/elastic/go-licenser
[2021-07-30T12:32:39.551Z] + go get -u golang.org/x/tools/cmd/goimports
[2021-07-30T12:32:46.504Z] + go get -u github.com/jstemmer/go-junit-report
[2021-07-30T12:32:47.187Z] Running in /var/lib/jenkins/workspace/_manager_package-registry_PR-703/src/github.com/elastic/package-registry
[2021-07-30T12:32:47.496Z] + mage -debug check
[2021-07-30T12:32:47.496Z] DEBUG: 12:32:47.247070 getting all non-mage files in .
[2021-07-30T12:32:48.063Z] DEBUG: 12:32:47.859251 found non-mage files artifacts.go||categories.go||favicon.go||handler.go||index.go||main.go||mime_types.go||package_index.go||search.go||static.go
[2021-07-30T12:32:48.063Z] DEBUG: 12:32:47.859278 marked file as non-mage: "artifacts.go"
[2021-07-30T12:32:48.063Z] DEBUG: 12:32:47.859283 marked file as non-mage: "categories.go"
[2021-07-30T12:32:48.063Z] DEBUG: 12:32:47.859287 marked file as non-mage: "favicon.go"
[2021-07-30T12:32:48.063Z] DEBUG: 12:32:47.859290 marked file as non-mage: "handler.go"
[2021-07-30T12:32:48.063Z] DEBUG: 12:32:47.859293 marked file as non-mage: "index.go"
[2021-07-30T12:32:48.063Z] DEBUG: 12:32:47.859296 marked file as non-mage: "main.go"
[2021-07-30T12:32:48.063Z] DEBUG: 12:32:47.859300 marked file as non-mage: "mime_types.go"
[2021-07-30T12:32:48.063Z] DEBUG: 12:32:47.859303 marked file as non-mage: "package_index.go"
[2021-07-30T12:32:48.063Z] DEBUG: 12:32:47.859306 marked file as non-mage: "search.go"
[2021-07-30T12:32:48.063Z] DEBUG: 12:32:47.859321 marked file as non-mage: "static.go"
[2021-07-30T12:32:48.063Z] DEBUG: 12:32:47.859326 getting all files plus mage files
[2021-07-30T12:32:48.630Z] DEBUG: 12:32:48.587313 time to scan for Magefiles: 1.34026429s
[2021-07-30T12:32:48.630Z] DEBUG: 12:32:48.587339 found magefiles: magefile.go
[2021-07-30T12:32:48.630Z] DEBUG: 12:32:48.592081 output exe is  /var/lib/jenkins/workspace/_manager_package-registry_PR-703/.magefile/f1b38259379e8e415e65aedbfd562ee28614ceb8
[2021-07-30T12:32:48.630Z] DEBUG: 12:32:48.597423 go build cache exists, will ignore any compiled binary
[2021-07-30T12:32:48.630Z] DEBUG: 12:32:48.597448 parsing files
[2021-07-30T12:32:48.630Z] DEBUG: 12:32:48.598130 found target AddLicenseHeaders
[2021-07-30T12:32:48.630Z] DEBUG: 12:32:48.598159 found target Build
[2021-07-30T12:32:48.630Z] DEBUG: 12:32:48.598167 found target Check
[2021-07-30T12:32:48.630Z] DEBUG: 12:32:48.598173 found target Clean
[2021-07-30T12:32:48.630Z] DEBUG: 12:32:48.598179 found target FetchPackageStorage
[2021-07-30T12:32:48.630Z] DEBUG: 12:32:48.598186 skipping function with invalid signature func FindFilesRecursive: ETOOMANYRETURNS
[2021-07-30T12:32:48.630Z] DEBUG: 12:32:48.598192 found target Format
[2021-07-30T12:32:48.630Z] DEBUG: 12:32:48.598221 found target GoImports
[2021-07-30T12:32:48.630Z] DEBUG: 12:32:48.598241 found target ModTidy
[2021-07-30T12:32:48.630Z] DEBUG: 12:32:48.598253 found target Test
[2021-07-30T12:32:48.630Z] DEBUG: 12:32:48.598277 time parse Magefiles: 821.43µs
[2021-07-30T12:32:48.630Z] DEBUG: 12:32:48.598295 Creating mainfile at mage_output_file.go
[2021-07-30T12:32:48.630Z] DEBUG: 12:32:48.598331 writing new file at mage_output_file.go
[2021-07-30T12:32:48.630Z] DEBUG: 12:32:48.599536 compiling to /var/lib/jenkins/workspace/_manager_package-registry_PR-703/.magefile/f1b38259379e8e415e65aedbfd562ee28614ceb8
[2021-07-30T12:32:48.630Z] DEBUG: 12:32:48.599540 compiling using gocmd: go
[2021-07-30T12:32:48.890Z] DEBUG: 12:32:48.661563 running go build -o /var/lib/jenkins/workspace/_manager_package-registry_PR-703/.magefile/f1b38259379e8e415e65aedbfd562ee28614ceb8 magefile.go mage_output_file.go
[2021-07-30T12:32:49.149Z] DEBUG: 12:32:49.131463 time to compile Magefile: 469.851084ms
[2021-07-30T12:32:49.149Z] DEBUG: 12:32:49.131525 running binary /var/lib/jenkins/workspace/_manager_package-registry_PR-703/.magefile/f1b38259379e8e415e65aedbfd562ee28614ceb8
[2021-07-30T12:32:49.149Z] DEBUG: 12:32:49.131550 running magefile with mage vars:
[2021-07-30T12:32:49.149Z] MAGEFILE_DEBUG=1
[2021-07-30T12:32:49.149Z] MAGEFILE_GOCMD=go
[2021-07-30T12:32:49.149Z] >> fmt - go-licenser: Adding missing headers
[2021-07-30T12:32:49.407Z] >> fmt - goimports: Formatting Go code
[2021-07-30T12:32:49.407Z] Cloning into 'build/package-storage'...
[2021-07-30T12:32:55.661Z] Updating files:  18% (5019/26793)
Updating files:  19% (5091/26793)
Updating files:  20% (5359/26793)
Updating files:  21% (5627/26793)
Updating files:  22% (5895/26793)
Updating files:  23% (6163/26793)
Updating files:  24% (6431/26793)
Updating files:  25% (6699/26793)
Updating files:  26% (6967/26793)
Updating files:  27% (7235/26793)
Updating files:  28% (7503/26793)
Updating files:  29% (7770/26793)
Updating files:  30% (8038/26793)
Updating files:  31% (8306/26793)
Updating files:  32% (8574/26793)
Updating files:  33% (8842/26793)
Updating files:  34% (9110/26793)
Updating files:  35% (9378/26793)
Updating files:  36% (9646/26793)
Updating files:  37% (9914/26793)
Updating files:  38% (10182/26793)
Updating files:  39% (10450/26793)
Updating files:  40% (10718/26793)
Updating files:  41% (10986/26793)
Updating files:  42% (11254/26793)
Updating files:  43% (11521/26793)
Updating files:  44% (11789/26793)
Updating files:  45% (12057/26793)
Updating files:  46% (12325/26793)
Updating files:  47% (12593/26793)
Updating files:  48% (12861/26793)
Updating files:  49% (13129/26793)
Updating files:  49% (13343/26793)
Updating files:  50% (13397/26793)
Updating files:  51% (13665/26793)
Updating files:  52% (13933/26793)
Updating files:  53% (14201/26793)
Updating files:  54% (14469/26793)
Updating files:  55% (14737/26793)
Updating files:  56% (15005/26793)
Updating files:  57% (15273/26793)
Updating files:  58% (15540/26793)
Updating files:  59% (15808/26793)
Updating files:  60% (16076/26793)
Updating files:  61% (16344/26793)
Updating files:  62% (16612/26793)
Updating files:  63% (16880/26793)
Updating files:  64% (17148/26793)
Updating files:  65% (17416/26793)
Updating files:  66% (17684/26793)
Updating files:  67% (17952/26793)
Updating files:  68% (18220/26793)
Updating files:  69% (18488/26793)
Updating files:  70% (18756/26793)
Updating files:  71% (19024/26793)
Updating files:  72% (19291/26793)
Updating files:  73% (19559/26793)
Updating files:  74% (19827/26793)
Updating files:  75% (20095/26793)
Updating files:  76% (20363/26793)
Updating files:  77% (20631/26793)
Updating files:  78% (20899/26793)
Updating files:  79% (21167/26793)
Updating files:  80% (21435/26793)
Updating files:  81% (21703/26793)
Updating files:  82% (21971/26793)
Updating files:  83% (22239/26793)
Updating files:  84% (22507/26793)
Updating files:  85% (22775/26793)
Updating files:  86% (23042/26793)
Updating files:  87% (23310/26793)
Updating files:  88% (23578/26793)
Updating files:  89% (23846/26793)
Updating files:  89% (24055/26793)
Updating files:  90% (24114/26793)
Updating files:  91% (24382/26793)
Updating files:  92% (24650/26793)
Updating files:  93% (24918/26793)
Updating files:  94% (25186/26793)
Updating files:  95% (25454/26793)
Updating files:  96% (25722/26793)
Updating files:  97% (25990/26793)
Updating files:  98% (26258/26793)
Updating files:  99% (26526/26793)
Updating files: 100% (26793/26793)
Updating files: 100% (26793/26793), done.
[2021-07-30T12:32:57.576Z] # github.com/elastic/package-registry/util
[2021-07-30T12:32:57.576Z] util/package.go:160:17: undefined: os.MkdirTemp
[2021-07-30T12:32:57.576Z] Error: running "go build ." failed with exit code 2
[2021-07-30T12:32:57.726Z] Stage "Build" skipped due to earlier failure(s)
[2021-07-30T12:32:57.789Z] Stage "Test" skipped due to earlier failure(s)
[2021-07-30T12:32:57.854Z] Stage "Publish Docker image" skipped due to earlier failure(s)
[2021-07-30T12:32:57.915Z] Stage "Publish PR Docker image" skipped due to earlier failure(s)
[2021-07-30T12:32:58.775Z] Running on Jenkins in /var/lib/jenkins/workspace/_manager_package-registry_PR-703
[2021-07-30T12:32:58.896Z] [INFO] getVaultSecret: Getting secrets
[2021-07-30T12:32:58.937Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2021-07-30T12:32:59.787Z] + chmod 755 generate-build-data.sh
[2021-07-30T12:32:59.788Z] + ./generate-build-data.sh https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Ingest-manager/package-registry/PR-703/ https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Ingest-manager/package-registry/PR-703/runs/1 FAILURE 153642
[2021-07-30T12:32:59.788Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Ingest-manager/package-registry/PR-703/runs/1/steps/?limit=10000 -o steps-info.json
[2021-07-30T12:32:59.788Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Ingest-manager/package-registry/PR-703/runs/1/tests/?status=FAILED -o tests-errors.json
[2021-07-30T12:33:00.038Z] Retry 1/3 exited 22, retrying in 1 seconds...
[2021-07-30T12:33:00.949Z] Retry 2/3 exited 22, retrying in 2 seconds...

```

</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

mtojek commented on util/fs.go:
```diff
@@ -0,0 +1,226 @@
+	case info.IsDir():
+		return NewExtractedPackageFileSystem(path)
+	case strings.HasSuffix(path, ".zip"):
+		return NewZipPackageFileSystem(path)
```
That reminded me: https://github.com/golang/tools/blob/master/godoc/vfs/zipfs/zipfs.go

jsoriano replied:
Oh, I was hoping to find something like this :slightly_smiling_face: With the new interfaces added in 1.16 these things should be more common.

mtojek reviewed (commented):
The idea of indexer seems to be the right direction, but it also shows the spaghetti code we have here. 

I understand that this will change the way we store packages in Package Storage - zip archives instead of raw files? Please keep in mind that we'll have to adjust our tooling to support this distribution form and also inform other parties about the initiative (teams contributing directly to Package Storage like endpoint or APM).

There are also some features we use thank to raw files:
* we can see differences between next package revisions ([PR changes](https://github.com/elastic/package-storage/commit/9fbb2f13fc114c9bdf85e02e640dcb9cfab9cb08))
* "grepping" is easy

I don't think it's impossible to plan replacements for them.

mtojek commented on main.go:
```diff
@@ -60,6 +60,10 @@ type Config struct {
 	CacheTimeCatchAll   time.Duration `config:"cache_time.catch_all"`
 }
 
+type Indexer interface {
+	GetPackages(context.Context, *util.GetPackagesOptions) (util.Packages, error)
+}
+
```
This is just a loose suggestion, I don't want to interfere or stop you from making the progress.

I guess you can extract all interfaces and domain classes to a separate package, so they can be used to implement more sophisticated Indexers and Storages (more or less similarly to https://github.com/elastic/beats/tree/master/libbeat/beat).

jsoriano replied:
I like to have the interfaces defined where they are used :slightly_smiling_face: Implementations don't really need it to implement it.

If we find that there may be more users of Indexers, we can reconsider and move the interface to a separate package (an alternative can also be to duplicate the definition, what can be fine in many cases too).

jsoriano commented:
@mtojek I am opening this for review.

I have updated the branch with recent changes related to categories and migrated the new code to follow the refactor, API tests pass now as they are in master.

I have also hidden the new feature behind an experimental flag, so we don't have surprises by trying to serve additional files. Still with this refactor we would have API and storage less coupled, and we would have the package filtering in a single place.

We can continue later with further optimizations or refactors as the one to use `io.FS` implementations, I have added them as postponed tasks in the description.

mtojek commented on static.go:
```diff
@@ -8,26 +8,56 @@ import (
+func staticHandler(indexer Indexer, cacheTime time.Duration) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		vars := mux.Vars(r)
+		packageName, ok := vars["packageName"]
+		if !ok {
+			badRequest(w, "missing package name")
 			return
 		}
+
+		packageVersion, ok := vars["packageVersion"]
+		if !ok {
+			badRequest(w, "missing package version")
+			return
+		}
+
+		_, err := semver.StrictNewVersion(packageVersion)
 		if err != nil {
-			log.Printf("stat package path '%s' failed: %v", r.URL.Path, err)
+			badRequest(w, "invalid package version")
+			return
+		}
 
+		fileName, ok := vars["name"]
+		if !ok {
+			badRequest(w, "missing file name")
+			return
+		}
```
nit: this can extracted to separate function for readability purposes

jsoriano replied:
Will add params handling to its own function.

mtojek commented on search.go:
```diff
@@ -80,179 +70,46 @@ func newSearchFilterFromParams(r *http.Request) (searchFilter, error) {
 	span, ctx := apm.StartSpan(ctx, "GetPackageOutput", "app")
 	defer span.End()
 
-	separator := "@"
```
I'm not sure if this logic wasn't placed here due to performance reasons. Maybe double-check it with @ruflin ?

jsoriano replied:
To what logic do you refer? To the whole search/filtering logic? I am moving it to a common place so all handlers have consistent behaviour, and we have a single place to look at to make performance optimizations if needed :slightly_smiling_face:

mtojek replied:
I meant the sorting logic below.

jsoriano replied:
Well, before it was sorting the packages received as values in a map of maps, in that case sorting the packages as is done here would imply to copy the packages first to a list and then copying them again when sorting, or creating a list of pointers to the packages and sorting it. In any of these cases sorting only the keys and then creating the list of packages was probably better.

Now this method receives already a list of pointers to the packages, what can be sorted right away without the need of building the list of keys to sort.

mtojek replied:
> Now this method receives already a list of pointers to the packages, what can be sorted right away without the need of building the list of keys to sort.

I was concerned that it operates on a list of relatively complex package structures. If this is not the thing, then feel free to resolve this comment.

mtojek commented on util/fs.go:
```diff
@@ -0,0 +1,207 @@
+	reader *zip.ReadCloser
+}
+
+func NewZipPackageFileSystem(p *Package) (*zipPackageFileSystem, error) {
```
Things are mixed here: exposed function with private structure.

mtojek commented on util/packages.go:
```diff
@@ -19,84 +20,326 @@ import (
+		paths:     paths,
+		label:     "FileSystemIndexer",
+		walkerFn:  walkerFn,
+		fsBuilder: fsBuilder,
```
maybe just:

```
fsBuilder: NewExtractedPackageFileSystem,
```

?

jsoriano replied:
`NewExtractedPackageFileSystem` returns the specific type `*ExtractedPackageFileSystem`, fsBuilder expects a builder that returns the `FileSystemBuilder` interface.

mtojek replied:
Ok, I see now. My intention was to get rid of the additional proxy function due to lang limitations. If it isn't possible, feel free to ignore this comment.

mtojek commented on util/packages.go:
```diff
@@ -19,84 +20,326 @@ import (
-			if err != nil {
-				return err
-			}
+	err := filepath.Walk(packagesPath, func(path string, info os.FileInfo, err error) error {
```
You can also consider WalkDir as it might be a bit more efficient for raw file system.

jsoriano replied:
Good idea, done.

mtojek reviewed (commented):
I left some comments, but nothing really breaking the idea. Good job!

I'm not sure if you have done it, but maybe you can try to build a Docker image with real content from package storage to check the feature (including compressed packages).

jsoriano commented on util/packages.go:
```diff
@@ -19,84 +20,326 @@ import (
+
+// NewFileSystemIndexer creates a new FileSystemIndexer for the given paths.
+func NewFileSystemIndexer(paths ...string) *FileSystemIndexer {
+	walkerFn := func(basePath, path string, info os.FileInfo, err error) error {
```
Current implementation doesn't complain if the directories don't exist, only if there are no packages, this behaviour is maintained here. We can reconsider this in a follow up.

jsoriano commented on util/packages.go:
```diff
@@ -19,84 +20,326 @@ import (
-// GetPackages returns a slice with all existing packages.
+// NewZipFileSystemIndexer creates a new ZipFileSystemIndexer for the given paths.
+func NewZipFileSystemIndexer(paths ...string) *FileSystemIndexer {
+	walkerFn := func(basePath, path string, info os.FileInfo, err error) error {
```
Added error handling for the walker errors. If a directory doesn't exist, it is ignored, to keep current behaviour.

jsoriano commented on util/packages.go:
```diff
@@ -5,6 +5,7 @@
 package util
```
Agree, this `util` package contains too many things, I will leave this for the end, or for a follow up :)

jsoriano commented on util/packages.go:
```diff
@@ -19,84 +20,326 @@ import (
-	if packageList != nil {
-		return packageList, nil
+func (i *FileSystemIndexer) GetPackages(ctx context.Context, opts *GetPackagesOptions) (Packages, error) {
+	if i.packageList == nil {
```
Adding an explicit init method for indexers.

jsoriano commented on util/fs.go:
```diff
@@ -0,0 +1,207 @@
+	return ioutil.ReadAll(f)
+}
+
+// virtualPackageFileSystem provide utils for package objects that don't correspond to
```
This is used now in tests, it can be used too as base for filesystems that instantiate packages without actually having them on disk.

jsoriano commented on util/fs.go:
```diff
@@ -0,0 +1,207 @@
+	path string
+}
+
+func NewExtractedPackageFileSystem(p *Package) (*extractedPackageFileSystem, error) {
```
Making type public.

jsoriano commented on main.go:
```diff
@@ -112,14 +115,23 @@ func initServer() *http.Server {
+	indexers := []Indexer{
+		util.NewFileSystemIndexer(packagesBasePaths...),
+	}
+	if serveZip {
```
I think this will be safe, I only added it last minute as a safety net to don't have it enabled by default in production by now :slightly_smiling_face: Not a strong opinion, so as you prefer :slightly_smiling_face:

jsoriano commented on util/fs.go:
```diff
@@ -0,0 +1,207 @@
+
+// Seek implements the seeker interface for zip files. This is inefficient, it shouldn't
+// be frequently used.
+func (f *zipFileSeeker) Seek(offset int64, whence int) (n int64, err error) {
```
Yes, this is hacky. Full story is that I wanted to use [`http.ServeContent`](https://pkg.go.dev/net/http#ServeContent) because it handles content type and so on, but this requires the file reader to implement `io.ReadSeeker`, an interface that zip file readers don't implement. So I added a minimal implementation for it.

I can give a thought to serve the file with the proper content type without `http.ServeContent`, what wouldn't require this hack, but would require to duplicat some things that this helper does.

mtojek replied:
Ok, let's leave it for now. Thank you for the explanation.

jsoriano commented on util/packages.go:
```diff
@@ -19,84 +20,326 @@ import (
-			}
+	err := filepath.Walk(packagesPath, func(path string, info os.FileInfo, err error) error {
+		err = i.walkerFn(packagesPath, path, info, err)
+		if err != walkerIndexFile {
```
Better, yes :+1:

ruflin commented:
I really like the idea of introducing the Indexer especially as it also gets rid of all the passing around of paths. It also means assuming at one point we need to have an Indexer as a service (ES?)) it can be added to the list.

I only skimmed through the changes and it contains a lot of refactoring too which is great. As you know, I'm not too big of a fan of large PRs and if possible it would be nice to take out some parts of the refactoring / cleanup into separate PRs as it will make quicker to review and later in case of issues figuring out when it happened. But as I'm not the main reviewer leave it to @mtojek and you to decide if you want to get it all in at once.
